chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,830 @@
+	if err != nil {
+		log.Errorf("error tracking CID in hotstore: %s; falling back to coldstore", err)
+		return s.cold.Put(blk)
+	}
+
+	return s.hot.Put(blk)
+}
+
+func (s *SplitStore) PutMany(blks []blocks.Block) error {
+	s.mx.Lock()
+	if s.curTs == nil {
+		s.mx.Unlock()
+		return s.cold.PutMany(blks)
+	}
+
+	epoch := s.curTs.Height()
+	s.mx.Unlock()
+
+	batch := make([]cid.Cid, 0, len(blks))
+	for _, blk := range blks {
+		batch = append(batch, blk.Cid())
+	}
+
+	err := s.snoop.PutBatch(batch, epoch)
+	if err != nil {
+		log.Errorf("error tracking CIDs in hotstore: %s; falling back to coldstore", err)
+		return s.cold.PutMany(blks)
+	}
+
+	return s.hot.PutMany(blks)
+}
+
+func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	chHot, err := s.hot.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+		defer close(ch)
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (s *SplitStore) HashOnRead(enabled bool) {
+	s.hot.HashOnRead(enabled)
+	s.cold.HashOnRead(enabled)
+}
+
+func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	err := s.hot.View(cid, cb)
+	switch err {
+	case bstore.ErrNotFound:
+		return s.cold.View(cid, cb)
+
+	default:
+		return err
+	}
+}
+
+// State tracking
+func (s *SplitStore) Start(cs *store.ChainStore) error {
+	s.cs = cs
+	s.curTs = cs.GetHeaviestTipSet()
+
+	// load base epoch from metadata ds
+	// if none, then use current epoch because it's a fresh start
+	bs, err := s.ds.Get(baseEpochKey)
+	switch err {
+	case nil:
+		s.baseEpoch = bytesToEpoch(bs)
+
+	case dstore.ErrNotFound:
+		if s.curTs == nil {
+			// this can happen in some tests
+			break
+		}
+
+		err = s.setBaseEpoch(s.curTs.Height())
+		if err != nil {
+			return err
+		}
+
+	default:
+		return err
+	}
+
+	// watch the chain
+	cs.SubscribeHeadChanges(s.HeadChange)
+
+	return nil
+}
+
+func (s *SplitStore) Close() error {
+	if atomic.LoadInt32(&s.compacting) == 1 {
+		log.Warn("ongoing compaction; waiting for it to finish...")
+		for atomic.LoadInt32(&s.compacting) == 1 {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return s.env.Close()
+}
+
+func (s *SplitStore) HeadChange(revert, apply []*types.TipSet) error {
+	s.mx.Lock()
+	s.curTs = apply[len(apply)-1]
+	epoch := s.curTs.Height()
+	s.mx.Unlock()
+
+	if !atomic.CompareAndSwapInt32(&s.compacting, 0, 1) {
+		// we are currently compacting, do nothing and wait for the next head change
+		return nil
+	}
+
+	if epoch-s.baseEpoch > CompactionThreshold {
+		go func() {
+			defer atomic.StoreInt32(&s.compacting, 0)
+
+			log.Info("compacting splitstore")
+			start := time.Now()
+
+			s.compact()
```

Comment:
We should pass the current epoch here to avoid races. It could happen that by the time the compaction runs and re-reads the current epoch, it has slid.

Reply:
ok, we can do that -- we should probably pass the tipset itself.

Reply:
done.